tutorial/tests/testsuite.py

```diff
@@ -255,6 +94,39 @@ def ipytest(self, line: str, cell: str):
             if not functions_to_run:
                 raise ValueError("No function to test defined in the cell")
 
+            # store execution count information for each cell
```

Comment:
I suggest to put the extraction of all definitions and function calls in an external function and call it here. It would make the code more legible